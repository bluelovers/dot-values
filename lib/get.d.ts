/**
 * @param  {Object} target
 * @param  {String} path
 * @param  {*}      value
 * @return {*}
 */
export declare function get<T, D = T>(target: object, path: string | string[], defaultValue?: D): T | D | (T | D)[];
export default get;
