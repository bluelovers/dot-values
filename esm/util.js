/// <reference path="../module.d.ts" />
import _get from 'get-value';
// @ts-ignore
import escapeSplit from 'escape-split';
const SymStar = Symbol('*');
const splitSpecific = escapeSplit({ delimiter: '.', escaper: '\\' });
export function r(fn, target, paths, extra, result) {
    paths = p(paths);
    return r0(fn, target, paths, extra, result);
}
export function r0(fn, target, paths, extra, result) {
    let symIndex = paths.indexOf(SymStar);
    if (symIndex < 0) {
        // @ts-ignore
        let value = fn(target, paths, extra);
        if (typeof result != 'undefined') {
            result.push(value);
        }
        return value;
    }
    result = result || [];
    if (symIndex === 0) {
        let keys = ObjectKeysExtra(target, {
            //nonEnumerable: true,
            symbol: true,
        });
        paths.shift();
        for (let k of keys) {
            r0(fn, target, [k, ...paths], extra, result);
        }
        return result;
    }
    else {
        let p1 = paths.slice(0, symIndex);
        let o = _get(target, p1);
        let p2 = paths.slice(symIndex);
        return r0(fn, o, p2, extra, result);
    }
}
export function ObjectKeysExtra(obj, options = {}) {
    let keys = options.nonEnumerable ?
        Object.getOwnPropertyNames(obj)
        : Object.keys(obj);
    if (options.symbol) {
        keys.push(...Object.getOwnPropertySymbols(obj));
    }
    return keys;
}
/**
 * handle input to paths
 *
 * @param {string | ITSKeys[]} paths
 * @returns {ITSKeys[]}
 */
export function p(paths) {
    return p1(p0(paths));
}
export function p0(paths) {
    if (Array.isArray(paths))
        return paths.slice();
    if (typeof paths === 'symbol')
        return [paths];
    return splitSpecific(paths);
}
export function p1(paths) {
    return paths
        .map(v => {
        if (v === '*') {
            return SymStar;
        }
        else if (v === '\\*') {
            v = '*';
        }
        return v;
    });
}
/**
 * check all value is true
 *
 * @param {boolean | boolean[]} result
 * @returns {boolean}
 */
export function b(result) {
    return Array.isArray(result)
        ? result.every(boolean => !!boolean)
        : result;
}
export function b_some(result) {
    return Array.isArray(result)
        ? result.some(boolean => !!boolean)
        : result;
}
//# sourceMappingURL=util.js.map