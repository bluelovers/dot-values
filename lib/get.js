"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.get = void 0;
const tslib_1 = require("tslib");
const util_1 = require("./util");
const get_value_1 = (0, tslib_1.__importDefault)(require("get-value"));
function get(target, paths, defaultValue) {
    return (0, util_1.r)(get_value_1.default, target, paths, { default: defaultValue });
}
exports.get = get;
exports.default = get;
//# sourceMappingURL=get.js.map