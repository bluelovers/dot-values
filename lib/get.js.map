{"version":3,"file":"get.js","sourceRoot":"","sources":["get.ts"],"names":[],"mappings":";;AAAA,iCAA8B;AAC9B,yCAA6B;AAE7B;;;;;GAKG;AACH,SAAgB,GAAG,CAAW,MAAc,EAAE,IAAuB,EAAE,YAAgB;IAEtF,OAAO,QAAC,CAAQ,mBAAI,EAAE,MAAM,EAAE,QAAC,CAAC,IAAI,CAAC,EAAE,EAAE,OAAO,EAAE,YAAY,EAAE,CAAC,CAAA;AAClE,CAAC;AAHD,kBAGC;AAED,kBAAe,GAAG,CAAA","sourcesContent":["import { r, p } from './util';\nimport _get from 'get-value';\n\n/**\n * @param  {Object} target\n * @param  {String} path\n * @param  {*}      value\n * @return {*}\n */\nexport function get<T, D = T>(target: object, path: string | string[], defaultValue?: D)\n{\n\treturn r<D | T>(_get, target, p(path), { default: defaultValue })\n}\n\nexport default get\n"]}