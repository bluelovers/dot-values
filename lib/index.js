"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteValue = exports.delete = exports.unsetValue = exports.hasValue = exports.setValue = exports.getValue = exports.unset = exports.has = exports.set = exports.get = void 0;
const tslib_1 = require("tslib");
const get_1 = (0, tslib_1.__importDefault)(require("./get"));
exports.get = get_1.default;
exports.getValue = get_1.default;
const set_1 = (0, tslib_1.__importDefault)(require("./set"));
exports.set = set_1.default;
exports.setValue = set_1.default;
const has_1 = (0, tslib_1.__importDefault)(require("./has"));
exports.has = has_1.default;
exports.hasValue = has_1.default;
const unset_1 = (0, tslib_1.__importDefault)(require("./unset"));
exports.unset = unset_1.default;
exports.unsetValue = unset_1.default;
exports.delete = unset_1.default;
exports.deleteValue = unset_1.default;
exports.default = {
    get: get_1.default,
    set: set_1.default,
    has: has_1.default,
    unset: unset_1.default,
    getValue: get_1.default,
    setValue: set_1.default,
    hasValue: has_1.default,
    unsetValue: unset_1.default,
    delete: unset_1.default,
    deleteValue: unset_1.default,
};
//# sourceMappingURL=index.js.map