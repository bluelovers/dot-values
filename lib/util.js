"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.b_some = exports.b = exports.p1 = exports.p0 = exports.p = exports.ObjectKeysExtra = exports.r0 = exports.r = void 0;
const tslib_1 = require("tslib");
/// <reference path="../module.d.ts" />
const get_value_1 = (0, tslib_1.__importDefault)(require("get-value"));
// @ts-ignore
const escape_split_1 = (0, tslib_1.__importDefault)(require("escape-split"));
const SymStar = Symbol('*');
const splitSpecific = (0, escape_split_1.default)({ delimiter: '.', escaper: '\\' });
function r(fn, target, paths, extra, result) {
    paths = p(paths);
    return r0(fn, target, paths, extra, result);
}
exports.r = r;
function r0(fn, target, paths, extra, result) {
    let symIndex = paths.indexOf(SymStar);
    if (symIndex < 0) {
        // @ts-ignore
        let value = fn(target, paths, extra);
        if (typeof result != 'undefined') {
            result.push(value);
        }
        return value;
    }
    result = result || [];
    if (symIndex === 0) {
        let keys = ObjectKeysExtra(target, {
            //nonEnumerable: true,
            symbol: true,
        });
        paths.shift();
        for (let k of keys) {
            r0(fn, target, [k, ...paths], extra, result);
        }
        return result;
    }
    else {
        let p1 = paths.slice(0, symIndex);
        let o = (0, get_value_1.default)(target, p1);
        let p2 = paths.slice(symIndex);
        return r0(fn, o, p2, extra, result);
    }
}
exports.r0 = r0;
function ObjectKeysExtra(obj, options = {}) {
    let keys = options.nonEnumerable ?
        Object.getOwnPropertyNames(obj)
        : Object.keys(obj);
    if (options.symbol) {
        keys.push(...Object.getOwnPropertySymbols(obj));
    }
    return keys;
}
exports.ObjectKeysExtra = ObjectKeysExtra;
/**
 * handle input to paths
 *
 * @param {string | ITSKeys[]} paths
 * @returns {ITSKeys[]}
 */
function p(paths) {
    return p1(p0(paths));
}
exports.p = p;
function p0(paths) {
    if (Array.isArray(paths))
        return paths.slice();
    if (typeof paths === 'symbol')
        return [paths];
    return splitSpecific(paths);
}
exports.p0 = p0;
function p1(paths) {
    return paths
        .map(v => {
        if (v === '*') {
            return SymStar;
        }
        else if (v === '\\*') {
            v = '*';
        }
        return v;
    });
}
exports.p1 = p1;
/**
 * check all value is true
 *
 * @param {boolean | boolean[]} result
 * @returns {boolean}
 */
function b(result) {
    return Array.isArray(result)
        ? result.every(boolean => !!boolean)
        : result;
}
exports.b = b;
function b_some(result) {
    return Array.isArray(result)
        ? result.some(boolean => !!boolean)
        : result;
}
exports.b_some = b_some;
//# sourceMappingURL=util.js.map