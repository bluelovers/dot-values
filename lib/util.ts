import _get from 'get-value';
import _set from 'set-value';
import _has from 'has-value';
import _unset from 'unset-value';

/**
 * @param  {String} fn
 * @param  {Object} target
 * @param  {Array.<String>} path
 * @param  {*} extra
 * @param  {Array} result (optional)
 * @return {*}
 */
export function r<T>(fn: typeof _get | typeof _set | typeof _has | typeof _unset, target: unknown | unknown[], path: string[], extra?: unknown, result?: T[]): T | T[]
{
	let symIndex = path.indexOf('*'),
		hasResult = Array.isArray(result),
		isLast = !path.length

	if (symIndex < 0 || isLast)
	{
		let value = isLast
			? target
			// @ts-ignore
			: fn(target, path.join('.'), extra)

		if (hasResult)
		{
			result.push(value)
		}

		return value
	}

	result = hasResult ? result : []

	if (!(symIndex === 0 && Array.isArray(target)))
	{
		return r(fn, _get(target as unknown[], path.shift()), path, extra, result)
	}

	path.shift()
	target.forEach((target) =>
	{
		r(fn, target, Array.from(path), extra, result)
	})

	return result
}

/**
 * @param {String|Array} path
 * @return {Array.<String>}
 */
export function p(path: string | string[])
{
	if (Array.isArray(path)) return path

	return path.split(/\.*(\*)\.*/)
		.filter(segment => segment !== '')
}

/**
 * @param {Array|Boolean} result
 * @return {Boolean}
 */
export function b(result: boolean | boolean[])
{
	return Array.isArray(result)
		? !result.filter(boolean => !boolean).length
		: result
}
