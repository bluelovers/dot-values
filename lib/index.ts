import get from './get';
import set from './set';
import has from './has';
import unset from './unset';

export {
	get,
	set,
	has,
	unset,
}

export {
	get as getValue,
	set as setValue,
	has as hasValue,
	unset as unsetValue,
}

export default {
	get,
	set,
	has,
	unset,
}
