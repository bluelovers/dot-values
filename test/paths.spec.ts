import { p } from '../lib/util';

describe(`check supported path`, () =>
{

	test(`symbol`, () =>
	{
		const key = Symbol('foo');

		let actual = p(key);
		let expected = [key];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`array`, () =>
	{
		const key = ['b', 'c', 'd'];

		let actual = p(key);
		let expected = ['b', 'c', 'd'];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`b.c.d`, () =>
	{
		const key = 'b.c.d';

		let actual = p(key);
		let expected = ['b', 'c', 'd'];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`allow keys to be whitespace`, () =>
	{
		const key = 'a. .a';

		let actual = p(key);
		let expected = ['a', ' ', 'a'];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`a\\.b.c.d.e`, () =>
	{
		const key = 'a\\.b.c.d.e';

		let actual = p(key);
		let expected = [
			"a.b",
			"c",
			"d",
			"e"
		];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`e\\.f\\.g`, () =>
	{
		const key = 'e\\.f\\.g';

		let actual = p(key);
		let expected = [
			"e.f.g"
		];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`a.{b.c}.{d.e}`, () =>
	{
		const key = 'a.{b.c}.{d.e}';

		let actual = p(key);
		let expected = [ 'a', '{b', 'c}', '{d', 'e}' ];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

})
