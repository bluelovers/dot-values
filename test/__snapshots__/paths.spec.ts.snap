// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check supported path a.{b.c}.{d.e} 1`] = `
Array [
  "a",
  "{b",
  "c}",
  "{d",
  "e}",
]
`;

exports[`check supported path a\\.b.c.d.e 1`] = `
Array [
  "a.b",
  "c",
  "d",
  "e",
]
`;

exports[`check supported path allow keys to be whitespace 1`] = `
Array [
  "a",
  " ",
  "a",
]
`;

exports[`check supported path array 1`] = `
Array [
  "b",
  "c",
  "d",
]
`;

exports[`check supported path b.c.d 1`] = `
Array [
  "b",
  "c",
  "d",
]
`;

exports[`check supported path e\\.f\\.g 1`] = `
Array [
  "e.f.g",
]
`;

exports[`check supported path symbol 1`] = `
Array [
  Symbol(foo),
]
`;
